\name{formatPlot}
\alias{formatPlot}
%- Also NEED an '\alias' for EACH other topic documented here.
\title{
formatPlot
}
\description{
plot realigned reads
}
\usage{
formatPlot(dftailA, mismatchcolor = "", title = "Alignment Plot", covMaxDisplay = Inf)
}
\arguments{
  \item{dftailA}{
    data frame containing output from alignView or ViewAndScore
  }
  \item{mismatchcolor}{
    color of mismatching bases. If unset each mismatched base will be a
    different color
  }
  \item{title}{
    title of plot
  }
  \item{covMaxDisplay}{
    The maximum number of readpairs allowed in the display
  }
  
}
\details{
  see example
}
\value{
  a ggplot2 object
}
\author{
  Eitan Halper-Stromberg
}
\examples{
path <- system.file("extdata", package="targetSeqView")
filename=file.path(path, "twoSVJunctionsFailed.txt")
nodes=1
registerDoMC(nodes)
retfail=ViewAndScore(filename=filename,bamFilePath=path,
estimateIndelRate=FALSE,estimateMmRate=FALSE,getReadLength=FALSE,
build='hg19',verbose=TRUE)
print('The likelihood scores for the events failing validation:')
print(retfail[[2]])
for(ii in 1:length(retfail[[1]])){
    p1=formatPlot(retfail[[1]][[ii]][[1]][[2]],title='Alignment supporting a structural variant')
    p2=formatPlot(retfail[[1]][[ii]][[2]][[2]],title='Alignment supporting no structural variant')
    p3=formatPlot(retfail[[1]][[ii]][[3]][[2]],title='Alignment supporting no structural variant')
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(3, 1)))
    print(p1,vp = viewport(layout.pos.row = 1, layout.pos.col=1))
    print(p2,vp = viewport(layout.pos.row = 2, layout.pos.col=1))
    print(p3,vp = viewport(layout.pos.row = 3, layout.pos.col=1))
}
}